gtsam/geometry/Cal3Fisheye.cpp

```diff
@@ -0,0 +1,216 @@
+Matrix3 Cal3Fisheye::K() const {
+  Matrix3 K;
+  K << fx_, s_, u0_, 0.0, fy_, v0_, 0.0, 0.0, 1.0;
+  return K;
+}
+
+/* ************************************************************************* */
+static Matrix29 D2dcalibration(const double xd, const double yd,
+                               const double xi, const double yi,
+                               const double t3, const double t5,
+                               const double t7, const double t9, const double r,
+                               Matrix2& DK) {
+  // order: fx, fy, s, u0, v0
+  Matrix25 DR1;
+  DR1 << xd, 0.0, yd, 1.0, 0.0, 0.0, yd, 0.0, 0.0, 1.0;
+
+  // order: k1, k2, k3, k4
+  Matrix24 DR2;
+  DR2 << t3 * xi, t5 * xi, t7 * xi, t9 * xi, t3 * yi, t5 * yi, t7 * yi, t9 * yi;
+  DR2 /= r;
+  Matrix29 D;
+  D << DR1, DK * DR2;
+  return D;
+}
+
+/* ************************************************************************* */
+static Matrix2 D2dintrinsic(const double xi, const double yi, const double r,
+                            const double td, const double t, const double tt,
+                            const double t4, const double t6, const double t8,
+                            const double k1, const double k2, const double k3,
+                            const double k4, const Matrix2& DK) {
+  const double dr_dxi = xi / sqrt(xi * xi + yi * yi);
+  const double dr_dyi = yi / sqrt(xi * xi + yi * yi);
+  const double dt_dr = 1 / (1 + r * r);
+  const double dtd_dt =
+      1 + 3 * k1 * tt + 5 * k2 * t4 + 7 * k3 * t6 + 9 * k4 * t8;
+  const double dtd_dxi = dtd_dt * dt_dr * dr_dxi;
+  const double dtd_dyi = dtd_dt * dt_dr * dr_dyi;
+
+  const double rinv = 1 / r;
+  const double rrinv = 1 / (r * r);
+  const double dxd_dxi =
+      dtd_dxi * xi * rinv + td * rinv + td * xi * (-rrinv) * dr_dxi;
+  const double dxd_dyi = dtd_dyi * xi * rinv - td * xi * rrinv * dr_dyi;
+  const double dyd_dxi = dtd_dxi * yi * rinv - td * yi * rrinv * dr_dxi;
+  const double dyd_dyi =
+      dtd_dyi * yi * rinv + td * rinv + td * yi * (-rrinv) * dr_dyi;
+
+  Matrix2 DR;
+  DR << dxd_dxi, dxd_dyi, dyd_dxi, dyd_dyi;
+
+  return DK * DR;
+}
+
+/* ************************************************************************* */
+Point2 Cal3Fisheye::uncalibrate(const Point2& p, OptionalJacobian<2, 9> H1,
+                                OptionalJacobian<2, 2> H2) const {
+  const double xi = p.x(), yi = p.y();
+  const double r = sqrt(xi * xi + yi * yi);
+  const double t = atan(r);
+  const double tt = t * t, t4 = tt * tt, t6 = tt * t4, t8 = t4 * t4;
+  const double td = t * (1 + k1_ * tt + k2_ * t4 + k3_ * t6 + k4_ * t8);
+  const double xd = td / r * xi, yd = td / r * yi;
+  Point2 uv(fx_ * xd + s_ * yd + u0_, fy_ * yd + v0_);
+
+  Matrix2 DK;
+  if (H1 || H2) DK << fx_, s_, 0.0, fy_;
+
+  // Derivative for calibration parameters (2 by 9)
+  if (H1)
+    *H1 = D2dcalibration(xd, yd, xi, yi, t * tt, t * t4, t * t6, t * t8, r, DK);
```

Comment:
If we don't need the separate Jacobian methods I'd inline them here, which would eliminate a lot of copy/paste, and will also allow to identify which variables are needed for evaluation only, and which *extra* things are needed for the derivatives.